crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -498,11 +485,37 @@ fn encode_settlement(
         )?);
     }
 
-    let pre_interactions = [
-        verification.pre_interactions.clone(),
-        trade.pre_interactions(),
-    ]
-    .concat();
+    // Execute interaction to set up trade right before transfering funds.
+    // This interaction does nothing if the user-provided pre-interactions
+    // already set everything up (e.g. approvals, balances). That way we can
+    // correctly verify quotes with or without these user pre-interactions
+    // with helpful error messages.
+    let trade_setup_interaction = {
+        let sell_amount = match query.kind {
+            OrderKind::Sell => query.in_amount.get(),
+            OrderKind::Buy => *out_amount,
+        };
+        let solver = dummy_contract!(Solver, trade.solver());
+        let setup_step = solver.ensure_trade_preconditions(
+            verification.from,
+            settlement,
+            query.sell_token,
+            sell_amount,
+            native_token,
+            TradeVerifier::SPARDOSE,
+        );
+        Interaction {
+            target: solver.address(),
+            value: 0.into(),
+            data: setup_step.tx.data.unwrap().0,
```

Comment:
`unwrap` looks scary, but it looks like it is used everywhere with the TX data.

Reply:
Yeah, it's a bit unfortunate that we have to read the value from a builder that doesn't know we populated the value before. Will use an `.expect()` to signal that this is okay.